translator/trace/oc_to_otlp.go

```diff
@@ -0,0 +1,433 @@
+		Events:                 events,
+		DroppedEventsCount:     droppedEventCount,
+		Links:                  links,
+		DroppedLinksCount:      droppedLinkCount,
+		Status:                 ocStatusToOtlp(ocSpan.Status),
+		LocalChildSpanCount:    childSpanCount,
+	}
+
+	return otlpSpan
+}
+
+func ocStatusToOtlp(ocStatus *octrace.Status) *otlptrace.Status {
+	if ocStatus == nil {
+		return nil
+	}
+	return &otlptrace.Status{
+		Code:    otlptrace.Status_StatusCode(ocStatus.Code),
+		Message: ocStatus.Message,
+	}
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToOtlp(ocTracestate *octrace.Span_Tracestate) string {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return sb.String()
+}
+
+func ocAttrsToOtlp(ocAttrs *octrace.Span_Attributes) (otlpAttrs []*otlpcommon.AttributeKeyValue, droppedCount uint32) {
+	if ocAttrs == nil {
+		otlpAttrs = nil
+		droppedCount = 0
+		return
+	}
+
+	otlpAttrs = make([]*otlpcommon.AttributeKeyValue, 0, len(ocAttrs.AttributeMap))
+	for key, ocAttr := range ocAttrs.AttributeMap {
+
+		otlpAttr := &otlpcommon.AttributeKeyValue{Key: key}
+
+		switch attribValue := ocAttr.Value.(type) {
+		case *octrace.AttributeValue_StringValue:
+			otlpAttr.StringValue = truncableStringToStr(attribValue.StringValue)
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_STRING
+
+		case *octrace.AttributeValue_IntValue:
+			otlpAttr.IntValue = attribValue.IntValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_INT
+
+		case *octrace.AttributeValue_BoolValue:
+			otlpAttr.BoolValue = attribValue.BoolValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_BOOL
+
+		case *octrace.AttributeValue_DoubleValue:
+			otlpAttr.DoubleValue = attribValue.DoubleValue
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_DOUBLE
+
+		default:
+			str := "<Unknown OpenCensus Attribute>"
+			otlpAttr.StringValue = str
+			otlpAttr.Type = otlpcommon.AttributeKeyValue_STRING
+		}
+		otlpAttrs = append(otlpAttrs, otlpAttr)
+	}
+	droppedCount = uint32(ocAttrs.DroppedAttributesCount)
+	return
+}
+
+func ocSpanKindToOtlp(ocKind octrace.Span_SpanKind, ocAttrs *octrace.Span_Attributes) otlptrace.Span_SpanKind {
+	switch ocKind {
+	case octrace.Span_SERVER:
+		return otlptrace.Span_SERVER
+
+	case octrace.Span_CLIENT:
+		return otlptrace.Span_CLIENT
+
+	case octrace.Span_SPAN_KIND_UNSPECIFIED:
+		// Span kind field is unspecified, check if TagSpanKind attribute is set.
+		// This can happen if span kind had no equivalent in OC, so we could represent it in
+		// the SpanKind. In that case the span kind may be a special attribute TagSpanKind.
+		if ocAttrs != nil {
+			kindAttr := ocAttrs.AttributeMap[TagSpanKind]
+			if kindAttr != nil {
+				strVal := kindAttr.Value.(*octrace.AttributeValue_StringValue)
```

Comment:
this type assertion should check an the bool, othewise it risks a panic

Reply:
Good catch!

Reply:
Fixed.